### ISSUE TITLE
CA2248 triggers for generic enums that may be of same type

### ISSUE THREAD

**Issue by jnyrup** [dotnet/roslyn-analyzers#4180]
### Analyzer

**Diagnostic ID**: [CA2248](https://docs.microsoft.com/visualstudio/code-quality/ca2248)

### Describe the improvement

When two generic enums `T1` and `T2` _may_ be of same type CA2248 is reported as if they are _definitely_ of different types.

### Describe suggestions on how to achieve the rule

```cs
public class AllEnums<T1>
    where T1 : Enum
{
    private readonly T1 value;

    public AllEnums(T1 value)
    {
        this.value = value;
    }

    public bool HaveFlag<T2>(T2 flag)
        where T2 : struct, T1 =>
        value.GetType() == flag.GetType()  // this should guard properly against 'T1' and 'T2' being different
            && value.HasFlag(flag); // CA2248 The argument type, 'T2', must be the same as the enum type 'T1'
}
```

### Additional context

Greatly simplified example of how the `AllEnums` class is used
```cs
[Flags]
public enum SomeFlags
{
    A
}

[Flags]
public enum OtherFlags
{
    A
}

public static class Test
{
    public static void Main()
    {
        new AllEnums<SomeFlags>(SomeFlags.A).HaveFlag(SomeFlags.A); // returns 'true' as expected
        //new AllEnums<SomeFlags>(SomeFlags.A).HaveFlag(OtherFlags.A); // CS0315
        //new AllEnums<SomeFlags>(SomeFlags.A).HaveFlag((Enum)null); // CS0453

        new AllEnums<Enum>(SomeFlags.A).HaveFlag(SomeFlags.A); // returns 'true' as expected
        new AllEnums<Enum>(SomeFlags.A).HaveFlag(OtherFlags.A); // throws ArgumentsException without 'GetType()' check
        //new AllEnums<Enum>(SomeFlags.A).HaveFlag((Enum)null); // CS0453
    }
}
```

**Comment by mavasani** [comment]
@Evangelink

**Comment by Evangelink** [comment]
I didn't considered this case at all... I will take care to submit a PR (the analyzer is currently looking that both types are the same from a `ITypeSymbol` point of view).

**Comment by Evangelink** [comment]
I have looked at the code and the only thing I can easily do is to say that if one of the type is generic then I bail-out. There is another in-between solution which is to compare the constraints of the enum but that seems harder...

@mavasani What would you recommend?

**Comment by mavasani** [comment]
Yes, lets do the more simpler, restrictive fix.
